react/week3/todolist/src/components/TodoForm.js

```diff
@@ -0,0 +1,81 @@
+import React, { useState } from "react";
+
+export default function TodoForm(props) {
+  const { todos, setTodos, setMessage } = props;
+  const [formState, setFormState] = useState({
+    description: "",
+    deadline: ""
+  });
+
+  function isDescriptionValid(description) {
+    if (description !== "") {
+      return true;
+    } else {
+      setMessage("You need to give a description");
+      return false;
+    }
+  }
+
+  function isDateValid(date) {
+    if (Date.now() - Date.parse(date) < 0) {
+      return true;
+    } else {
+      setMessage("Deadline must be later than now.");
+      return false;
+    }
+  }
+
+  function handleFormChange(e) {
+    setFormState({ ...formState, [e.target.name]: e.target.value });
+  }
+
+  function addToDo(e) {
+    e.preventDefault();
+    if (
+      isDescriptionValid(formState.description) &&
+      isDateValid(formState.deadline)
+    ) {
+      const newTodo = {
+        id: `${Math.floor(Math.random() * 10000)}`,
```

Comment:
Generally try to avoid math random for the 'key'. 
https://reactjs.org/docs/reconciliation.html#:~:text=keys%20should%20be%20stable%2C%20predictable%2C%20and%20unique.%20unstable%20keys%20(like%20those%20produced%20by%20math.random())%20will%20cause%20many%20component%20instances%20and%20dom%20nodes%20to%20be%20unnecessarily%20recreated%2C%20which%20can%20cause%20performance%20degradation%20and%20lost%20state%20in%20child%20components.

An alternative would be nanoid.
Another alternative would be to keep track of the ids your current todos have, and when creating a new one, increment that id. Note this is different from simply using the list's index!